File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int>st;
        st.push(-1);
        unordered_map<int,int>ans;
         vector<int> result;
     
            for(int j=nums2.size()-1;j>=0;j--)
            {
                
                    int curr=nums2[j];
                    while(st.top()!=-1&&st.top()<=curr)
                    {
                        st.pop();
                    }
                    ans[nums2[j]]=st.top();
                    st.push(nums2[j]);
        
            }

        for(auto num:nums1)
        {
            result.push_back(ans[num]);
        }
        return result;

    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer>st=new Stack<>();
        st.push(-1);
        HashMap<Integer,Integer>map=new HashMap<>();
        

        for(int i=nums2.length-1;i>=0;i--)
        {
              int curr=nums2[i];
              while(st.peek()!=-1 && st.peek()<=curr)
              {
                st.pop();
              }
              map.put(curr ,st.peek());
              st.push(nums2[i]);
        }
        
        int [] result=new int[nums1.length];
        for( int i=0;i<nums1.length;i++)
        {
            result[i]=map.get(nums1[i]);
        }
        return result;
    }
}